code/modules/admin/banjob.dm

```diff
@@ -1,109 +1,21 @@
-//This file was auto-corrected by findeclaration.exe on 25.5.2012 20:42:32
-
-var/jobban_runonce			// Updates legacy bans with new info
-var/jobban_keylist[0]		//to store the keys & ranks
-
-/proc/jobban_fullban(mob/M, rank, reason)
-	if (!M || !M.key) return
-	jobban_keylist.Add(text("[M.ckey] - [rank] ## [reason]"))
-	jobban_savebanfile()
-
-/proc/jobban_client_fullban(ckey, rank)
-	if (!ckey || !rank) return
-	jobban_keylist.Add(text("[ckey] - [rank]"))
-	jobban_savebanfile()
-
 //returns a reason if M is banned from rank, returns 0 otherwise
 /proc/jobban_isbanned(mob/M, rank)
-	if(M && rank)
-		for (var/s in jobban_keylist)
-			if( findtext(s,"[M.ckey] - [rank]") == 1 )
-				var/startpos = findtext(s, "## ")+3
-				if(startpos && startpos<length(s))
-					var/text = copytext(s, startpos, 0)
-					if(text)
-						return text
-				return "Reason Unspecified"
+	if(!M || !M.client)
+		return 0
+	if(!M.client.jobbancache)
+		jobban_buildcache(M.client)
+	for(var/eachrank in M.client.jobbancache)
+		if(eachrank == rank)
+			return M.client.jobbancache[rank]
 	return 0
 
-/*
-DEBUG
-/mob/verb/list_all_jobbans()
-	set name = "list all jobbans"
-
-	for(var/s in jobban_keylist)
-		world << s
-
-/mob/verb/reload_jobbans()
-	set name = "reload jobbans"
-
-	jobban_loadbanfile()
-*/
-
-/proc/jobban_loadbanfile()
-	if(config.ban_legacy_system)
-		var/savefile/S=new("data/job_full.ban")
-		S["keys[0]"] >> jobban_keylist
-		log_admin("Loading jobban_rank")
-		S["runonce"] >> jobban_runonce
-
-		if (!length(jobban_keylist))
-			jobban_keylist=list()
-			log_admin("jobban_keylist was empty")
-	else
-		if(!establish_db_connection())
-			world.log << "Database connection failed. Reverting to the legacy ban system."
-			diary << "Database connection failed. Reverting to the legacy ban system."
-			config.ban_legacy_system = 1
-			jobban_loadbanfile()
-			return
-
-		//Job permabans
-		var/DBQuery/query = dbcon.NewQuery("SELECT ckey, job FROM [format_table_name("ban")] WHERE bantype = 'JOB_PERMABAN' AND isnull(unbanned)")
+/proc/jobban_buildcache(client/C)
+	if(C && istype(C))
+		C.jobbancache = list()
+		var/DBQuery/query = dbcon.NewQuery("SELECT job, reason FROM [format_table_name("ban")] WHERE ckey = '[C.ckey]' AND (bantype = 'JOB_PERMABAN'  OR (bantype = 'JOB_TEMPBAN' AND expiration_time > Now())) AND isnull(unbanned)")

```

Comment:
Yep, the query was more or less copy-pasted from the preferences code, with one slight tweak.